jerry-core/ecma/operations/ecma-container-object.c

```diff
@@ -614,12 +620,203 @@ ecma_op_container_delete (ecma_value_t this_arg, /**< this argument */
+                                                                  : ECMA_BUILTIN_ID_MAP_ITERATOR_PROTOTYPE),
+                                         is_set ? ECMA_PSEUDO_SET_ITERATOR : ECMA_PSEUDO_MAP_ITERATOR,
+                                         type);
+} /* ecma_op_container_create_iterator */
+
+/**
+ * The %{Set, Map}IteratorPrototype% object's 'next' routine
+ *
+ * See also:
+ *          ECMA-262 v6, 23.1.5.2.1
+ *          ECMA-262 v6, 23.2.5.2.1
+ *
+ * Note:
+ *     Returned value must be freed with ecma_free_value.
+ *
+ * @return iterator result object, if success
+ *         error - otherwise
+ */
+ecma_value_t
+ecma_op_container_iterator_next (ecma_value_t this_val, /**< this argument */
+                                 bool is_set) /**< true - to perform Set operations
+                                               *   false - to perform Map operations */
+{
+  if (!ecma_is_value_object (this_val))
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Argument 'this' is not an object."));
+  }
+
+  ecma_object_t *obj_p = ecma_get_object_from_value (this_val);
+  ecma_extended_object_t *ext_obj_p = (ecma_extended_object_t *) obj_p;
+
+  if (ecma_get_object_type (obj_p) == ECMA_OBJECT_TYPE_PSEUDO_ARRAY
+      && ext_obj_p->u.pseudo_array.type != (is_set ? ECMA_PSEUDO_SET_ITERATOR : ECMA_PSEUDO_MAP_ITERATOR))
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Argument 'this' is not an iterator."));
+  }
+
+  ecma_map_object_t *map_object_p = ECMA_GET_POINTER (ecma_map_object_t,
+                                                      ext_obj_p->u.pseudo_array.u2.iterated_value_cp);
+
+  if (map_object_p == NULL)
+  {
+    return ecma_create_iter_result_object (ECMA_VALUE_UNDEFINED, ECMA_VALUE_TRUE);
+  }
+
+  ecma_object_t *internal_obj_p = ecma_get_object_from_value (map_object_p->header.u.class_prop.u.value);
+  ecma_collection_header_t *props_p = ecma_op_object_get_property_names (internal_obj_p, ECMA_LIST_NO_OPTS);
+
+  uint32_t length = props_p->item_count;
+  uint32_t index = ext_obj_p->u.pseudo_array.u1.iterator_index;
+
+  if (JERRY_UNLIKELY (index == ECMA_ITERATOR_INDEX_LIMIT))
+  {
+    /* After the ECMA_ITERATOR_INDEX_LIMIT limit is reached the [[%Iterator%NextIndex]]
+       property is stored as an internal property */
+    ecma_string_t *prop_name_p = ecma_get_magic_string (LIT_INTERNAL_MAGIC_STRING_ITERATOR_NEXT_INDEX);
+    ecma_value_t index_value = ecma_op_object_get (obj_p, prop_name_p);
+
+    if (!ecma_is_value_undefined (index_value))
+    {
+      index = (uint32_t) (ecma_get_number_from_value (index_value) + 1);
+    }
+
+    ecma_value_t put_result = ecma_op_object_put (obj_p,
+                                                  prop_name_p,
+                                                  ecma_make_uint32_value (index),
+                                                  true);
+
+    JERRY_ASSERT (ecma_is_value_true (put_result));
```

Comment:
Are you sure an assert is enough here? It could be an error value as well which should be returned from the function.

Reply:
Good catch!